src/bin/handle-git-dir-change.ts

```diff
@@ -0,0 +1,221 @@
+type GitDiffStatus = 'A' | 'D' | 'M' | 'T' | 'U' | 'X' | string;
+
+class InstallationError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'InstallationError';
+  }
+}
+
+/**
+ * Get the absolute path of a relative path
+ * @param {string} dir the wildcard directory containing git change, not necessarily the root git directory
+ * @returns {string} the absolute path relative to the path that the user executed the bash command in
+ */
+export function resolvePath(dir: string) {
+  const absoluteDir = path.resolve(process.cwd(), dir);
+  return absoluteDir;
+}
+
+/**
+ * Get the git root directory.
+ * Errors if the directory provided is not a git directory.
+ * @param {string} dir an absolute directory
+ * @returns {string} the absolute path of the git directory root
+ */
+export function findRepoRoot(dir: string): string {
+  try {
+    return execSync('git rev-parse --show-toplevel', {cwd: dir})
+      .toString()
+      .trimRight(); // remove the trailing \n
+  } catch (err) {
+    logger.error(`The directory provided is not a git directory: ${dir}`);
+    throw err;
+  }
+}
+
+/**
+ * Returns the git diff old/new mode, status, and path. Given a git diff.
+ * Errors if there is a parsing error
+ * @param {string} gitDiffPattern A single file diff. Renames and copies are broken up into separate diffs. See https://git-scm.com/docs/git-diff#Documentation/git-diff.txt-git-diff-filesltpatterngt82308203 for more details
+ * @returns indexable git diff fields: old/new mode, status, and path
+ */
+function parseGitDiff(
+  gitDiffPattern: string
+): {
+  oldMode: FileMode;
+  newMode: FileMode;
+  status: GitDiffStatus;
+  relativePath: string;
+} {
+  try {
+    const fields: string[] = gitDiffPattern.split(' ');
+    const newMode = fields[1] as FileMode;
+    const oldMode = fields[0].substring(1) as FileMode;
+    const statusAndPath = fields[4].split('\t');
+    const status = statusAndPath[0] as GitDiffStatus;
+    const relativePath = statusAndPath[1];
+    return {oldMode, newMode, status, relativePath};
+  } catch (err) {
+    logger.warning(
+      `\`git diff --raw\` may have changed formats: \n ${gitDiffPattern}`
+    );
+    throw err;
+  }
+}
+
+/**
+ * Get the GitHub mode, file content, and relative path asynchronously
+ * Rejects if there is a git diff error, or if the file contents could not be loaded.
+ * @param {string} gitRootDir the root of the local GitHub repository
+ * @param {string} gitDiffPattern A single file diff. Renames and copies are broken up into separate diffs. See https://git-scm.com/docs/git-diff#Documentation/git-diff.txt-git-diff-filesltpatterngt82308203 for more details
+ * @returns {Promise<GitFileData>} the current mode, the relative path of the file in the Git Repository, and the file status.
+ */
+export function getGitFileData(
+  gitRootDir: string,
+  gitDiffPattern: string
+): Promise<GitFileData> {
+  return new Promise((resolve, reject) => {
```

Comment:
Would async/await be easier here?

Reply:
I felt it was easier to explicitly be able to get the `readfile` error for clearer error messaging. Otherwise I'd have to have nested try catch blocks if i wanted more precise error logging.